mlir/lib/Dialect/EmitC/IR/EmitC.cpp

```diff
@@ -954,6 +972,8 @@ LogicalResult emitc::ArrayType::verify(
   for (int64_t dim : shape) {
     if (dim < 0)
       return emitError() << "dimensions must have non-negative size";
+    if (dim == ShapedType::kDynamic)
+      return emitError() << "dimensions must have static size";
```

Comment:
We may revisit this later, since `expand_shape` brings dynamic shape sizes as operands, and such `reinterepret_casts` seem to be supported in C++.